{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = '../../ADA2017-Tutorials/02 - Intro to Pandas/Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "pd.options.mode.chained_assignment = None  # default='warn', Mutes warnings when copying a slice from a DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "pathSL = DATA_FOLDER + \"/ebola/sl_data\"\n",
    "allFilesSL = glob.glob(pathSL + \"/*.csv\")\n",
    "frameSL = pd.DataFrame()\n",
    "listSL = []\n",
    "for fileSL in allFilesSL:\n",
    "    df = pd.read_csv(fileSL,index_col=None, header=0)\n",
    "    listSL.append(df)\n",
    "frameSL = pd.concat(listSL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathLI = DATA_FOLDER + \"/ebola/liberia_data\"\n",
    "allFilesLI = glob.glob(pathLI + \"/*.csv\")\n",
    "frameLI = pd.DataFrame()\n",
    "listLI = []\n",
    "for fileLI in allFilesLI:\n",
    "    df = pd.read_csv(fileLI,index_col=None, header=0)\n",
    "    listLI.append(df)\n",
    "frameLI = pd.concat(listLI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pathGU = DATA_FOLDER + \"/ebola/guinea_data\"\n",
    "allFilesGU = glob.glob(pathGU + \"/*.csv\")\n",
    "frameGU = pd.DataFrame()\n",
    "listGU = []\n",
    "for fileGU in allFilesGU:\n",
    "    df = pd.read_csv(fileGU,index_col=None, header=0)\n",
    "    listGU.append(df)\n",
    "frameGU = pd.concat(listGU)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Variable</th>\n",
       "      <th>National</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"22\" valign=\"top\">Guinea</th>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-04</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-27</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-30</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-31</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-02</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-04</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-07</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-08</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-09</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-11</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-14</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-16</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-17</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-19</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-21</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-22</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-24</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-26</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-30</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-10-01</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"8\" valign=\"top\">Sierra Leone</th>\n",
       "      <th>4</th>\n",
       "      <td>2014-08-12</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2014-08-12</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-08-13</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2014-08-13</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-08-14</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2014-08-14</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-08-15</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2014-08-15</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">Liberia</th>\n",
       "      <th>4</th>\n",
       "      <td>2014-10-24 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-10-25 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-10-28 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-10-29 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-10-30 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-10-31 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-02 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-04 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-08 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-14 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-15 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-19 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-20 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-21 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-23 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-24 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-26 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-27 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-28 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-29 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-11-30 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-01 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-02 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-03 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-04 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-05 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-06 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-07 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-08 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2014-12-09 00:00:00</td>\n",
       "      <td>death_confirmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>329 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Date         Variable National\n",
       "Guinea       2            2014-08-04    new_confirmed        4\n",
       "             2            2014-08-26    new_confirmed       10\n",
       "             2            2014-08-27    new_confirmed       10\n",
       "             2            2014-08-30    new_confirmed        9\n",
       "             2            2014-08-31    new_confirmed       29\n",
       "             2            2014-09-02    new_confirmed       14\n",
       "             2            2014-09-04    new_confirmed       17\n",
       "             2            2014-09-07    new_confirmed       11\n",
       "             2            2014-09-08    new_confirmed       11\n",
       "             2            2014-09-09    new_confirmed        7\n",
       "             2            2014-09-11    new_confirmed       13\n",
       "             2            2014-09-14    new_confirmed       14\n",
       "             2            2014-09-16    new_confirmed        6\n",
       "             2            2014-09-17    new_confirmed        6\n",
       "             2            2014-09-19    new_confirmed       12\n",
       "             2            2014-09-21    new_confirmed       17\n",
       "             2            2014-09-22    new_confirmed       14\n",
       "             2            2014-09-23    new_confirmed       25\n",
       "             2            2014-09-24    new_confirmed       20\n",
       "             2            2014-09-26    new_confirmed       10\n",
       "             2            2014-09-30    new_confirmed       11\n",
       "             2            2014-10-01    new_confirmed        6\n",
       "Sierra Leone 4            2014-08-12    new_confirmed       11\n",
       "             11           2014-08-12  death_confirmed      264\n",
       "             4            2014-08-13    new_confirmed       15\n",
       "             11           2014-08-13  death_confirmed      273\n",
       "             4            2014-08-14    new_confirmed       13\n",
       "             11           2014-08-14  death_confirmed      280\n",
       "             4            2014-08-15    new_confirmed       10\n",
       "             11           2014-08-15  death_confirmed      287\n",
       "...                              ...              ...      ...\n",
       "Liberia      4   2014-10-24 00:00:00  death_confirmed      NaN\n",
       "             4   2014-10-25 00:00:00  death_confirmed      NaN\n",
       "             4   2014-10-28 00:00:00  death_confirmed      NaN\n",
       "             4   2014-10-29 00:00:00  death_confirmed      NaN\n",
       "             4   2014-10-30 00:00:00  death_confirmed      NaN\n",
       "             4   2014-10-31 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-02 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-04 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-08 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-14 00:00:00  death_confirmed        0\n",
       "             4   2014-11-15 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-19 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-20 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-21 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-23 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-24 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-26 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-27 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-28 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-29 00:00:00  death_confirmed      NaN\n",
       "             4   2014-11-30 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-01 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-02 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-03 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-04 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-05 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-06 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-07 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-08 00:00:00  death_confirmed      NaN\n",
       "             4   2014-12-09 00:00:00  death_confirmed      NaN\n",
       "\n",
       "[329 rows x 3 columns]"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameSL.rename(columns={'date':'Date', 'variable':'Variable'}, inplace=True)\n",
    "frameLI['Date'] = pd.to_datetime(frameLI.Date)\n",
    "frameGU.rename(columns={'Totals':'National','Description':'Variable' }, inplace=True)\n",
    "frameGU = frameGU[['Date', 'Variable', 'National']] \n",
    "frameSL = frameSL[['Date', 'Variable', 'National']]\n",
    "frameLI = frameLI[['Date', 'Variable', 'National']]\n",
    "frameTmp = [frameGU, frameSL, frameLI]\n",
    "frame = pd.concat(frameTmp, keys=['Guinea', 'Sierra Leone', 'Liberia'])\n",
    "new = ['New cases of confirmed','New case/s (confirmed)']\n",
    "deaths=['New deaths registered today (confirmed)','Total death/s in confirmed cases']\n",
    "for newi in new:\n",
    "    frame.Variable = frame.Variable.str.replace(newi, 'new_confirmed')\n",
    "for deathi in deaths:\n",
    "    frame.Variable = frame.Variable.str.replace(deathi, 'death_confirmed') some dont work?\n",
    "PhilterList = ['new_confirmed', 'death_confirmed']\n",
    "frame = frame.loc[frame['Variable'].isin(PhilterList)]\n",
    "frame.dropna(thresh=1)  how to include NaN\n",
    "frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['New cases of suspects', 'New cases of probables', 'new_confirmed',\n",
       "       'Total new cases registered so far', 'Total cases of suspects',\n",
       "       'Total cases of probables', 'Total cases of confirmed',\n",
       "       'Cumulative (confirmed + probable + suspects)',\n",
       "       'New deaths registered today',\n",
       "       'New deaths registered today (confirmed)',\n",
       "       'New deaths registered today (probables)',\n",
       "       'New deaths registered today (suspects)',\n",
       "       'Total deaths of suspects', 'Total deaths of probables',\n",
       "       'Total deaths of confirmed',\n",
       "       'Total deaths (confirmed + probables + suspects)',\n",
       "       'Total PEC center today', 'Total PEC center today (confirmed)',\n",
       "       'Total PEC center today (probables)',\n",
       "       'Total PEC center today (suspects)',\n",
       "       'Total of deaths in confirmed cases in CTE',\n",
       "       'Total of cured in confirmed cases in CTE',\n",
       "       'Number of male confirmed cases',\n",
       "       'Number of female confirmed cases',\n",
       "       'Number of male probables cases',\n",
       "       'Number of female probables cases', 'Number of male suspects cases',\n",
       "       'Number of female suspects cases', 'Total number of male cases',\n",
       "       'Total number of female cases',\n",
       "       'Number of confirmed cases among health workers',\n",
       "       'Number of deaths of confirmed cases among health workers',\n",
       "       'Number of probables cases among health workers',\n",
       "       'Number of deaths of probables cases among health workers',\n",
       "       'Number of suspects cases among health workers',\n",
       "       'Number of death of confirmed cases among health workers',\n",
       "       'Total contacts registered from start date',\n",
       "       'Number of contacts to follow today',\n",
       "       'Number of contacts out of track',\n",
       "       'Number of contacts followed yesterday',\n",
       "       'Number of patients tested', 'Number of samples collected',\n",
       "       'Total suspected non-class cases', 'New deaths registered',\n",
       "       'Fatality rate for confirmed and probables',\n",
       "       'new_confirmed among health workers',\n",
       "       'Total case of confirmed among health workers',\n",
       "       'New deaths registered among health workers',\n",
       "       'Total deaths registered among health workers',\n",
       "       'New admits to CTE so far',\n",
       "       'Total number of hospitalized cases in CTE',\n",
       "       'Total number of exits from CTE',\n",
       "       'Total number of admissions to CTE',\n",
       "       'New contacts registered so far',\n",
       "       'Number of contacts followed today',\n",
       "       'Number of contacts out of the track 21 days',\n",
       "       'Number of contacts lost to follow up',\n",
       "       'Number of samples collected today', 'Number of samples under test',\n",
       "       'Total samples tested', 'population', 'new_noncase',\n",
       "       'new_suspected', 'new_probable', 'cum_noncase', 'cum_suspected',\n",
       "       'cum_probable', 'cum_confirmed', 'death_suspected',\n",
       "       'death_probable', 'death_confirmed', 'cfr', 'etc_new_admission',\n",
       "       'etc_currently_admitted', 'etc_cum_admission', 'etc_new_deaths',\n",
       "       'etc_cum_deaths', 'etc_new_discharges', 'etc_cum_discharges',\n",
       "       'cum_contacts', 'cum_completed_contacts', 'contacts_followed',\n",
       "       'new_contacts', 'contacts_healthy', 'contacts_ill',\n",
       "       'contacts_not_seen', 'new_completed_contacts', 'percent_seen',\n",
       "       'total_lab_samples', 'repeat_samples', 'new_samples',\n",
       "       'new_positive', 'new_negative', 'pending', 'positive_corpse',\n",
       "       'negative_corpse', 'Specimens collected',\n",
       "       'Specimens pending for testing', 'Total specimens tested',\n",
       "       'Newly reported deaths', 'Total death/s in probable cases',\n",
       "       'Total death/s in suspected cases',\n",
       "       'Total death/s in confirmed, probable, suspected cases',\n",
       "       'Case Fatality Rate (CFR) - Confirmed & Probable Cases',\n",
       "       'Newly reported contacts', 'Total contacts listed',\n",
       "       'Currently under follow-up', 'Contacts seen',\n",
       "       'Contacts who completed 21 day follow-up',\n",
       "       'Contacts lost to follow-up', 'New admissions',\n",
       "       'Total no. currently in Treatment Units', 'Total discharges',\n",
       "       'Cumulative admission/isolation', 'Newly Reported Cases in HCW',\n",
       "       'Cumulative cases among HCW', 'Newly Reported deaths in HCW',\n",
       "       'Cumulative deaths among HCW', 'New Case/s (Suspected)',\n",
       "       'New Case/s (Probable)', 'New case/s (confirmed)',\n",
       "       'Total suspected cases', 'Total probable cases',\n",
       "       'Total confirmed cases',\n",
       "       'Total Number of Confirmed Cases of Sierra Leonean Nationality',\n",
       "       'Total Number of Confirmed Cases of Guinean Nationality',\n",
       "       'Cumulative (confirmed + probable + suspected)',\n",
       "       'Total Case/s (Suspected)', 'Total Case/s (Probable)',\n",
       "       'Total case/s (confirmed)', 'Cumulative CFR',\n",
       "       'Total death/s in confirmed,  probable, suspected cases',\n",
       "       'Cumulative confirmed, probable and suspected cases',\n",
       "       'Total death/s in confirmed, \\n probable, suspected cases',\n",
       "       'Case Fatality Rate (CFR) - \\n Confirmed & Probable Cases',\n",
       "       'Contacts who completed 21 day \\n follow-up',\n",
       "       'Total no. currently in Treatment \\n Units',\n",
       "       'Total Number of Confirmed Cases \\n of Sierra Leonean Nationality',\n",
       "       'Total Number of Confirmed Cases \\n of Guinean Nationality'], dtype=object)"
      ]
     },
     "execution_count": 306,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frame.Variable.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Beyla', 'Boffa', 'Conakry', 'Coyah', 'Dabola', 'Dalaba', 'Date',\n",
       "       'Variable', 'Dinguiraye', 'Dubreka', 'Forecariah', 'Gueckedou',\n",
       "       'Kerouane', 'Kindia', 'Kissidougou', 'Kouroussa', 'Lola', 'Macenta',\n",
       "       'Mzerekore', 'Nzerekore', 'Pita', 'Siguiri', 'Telimele', 'Totals',\n",
       "       'Yomou'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameGU.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frameGU.Variable = frameGU.Variable.str.replace('New cases of confirmed','new_confirmed')\n",
    "frameGU.Variable = frameGU.Variable.str.replace('New deaths registered today (confirmed)','death_confirmed')\n",
    "frameLI.Variable = frameLI.Variable.str.replace('New case/s (confirmed)','new_confirmed')\n",
    "frameLI.Variable = frameLI.Variable.str.replace('Total death/s in confirmed cases','death_confirmed')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Variable</th>\n",
       "      <th>Totals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-04</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-26</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-27</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-30</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-08-31</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-02</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-04</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-07</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-08</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-09</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-11</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-14</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-16</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-17</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-19</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-21</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-22</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-23</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-24</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-26</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-09-30</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2014-10-01</td>\n",
       "      <td>new_confirmed</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Date       Variable Totals\n",
       "2  2014-08-04  new_confirmed      4\n",
       "2  2014-08-26  new_confirmed     10\n",
       "2  2014-08-27  new_confirmed     10\n",
       "2  2014-08-30  new_confirmed      9\n",
       "2  2014-08-31  new_confirmed     29\n",
       "2  2014-09-02  new_confirmed     14\n",
       "2  2014-09-04  new_confirmed     17\n",
       "2  2014-09-07  new_confirmed     11\n",
       "2  2014-09-08  new_confirmed     11\n",
       "2  2014-09-09  new_confirmed      7\n",
       "2  2014-09-11  new_confirmed     13\n",
       "2  2014-09-14  new_confirmed     14\n",
       "2  2014-09-16  new_confirmed      6\n",
       "2  2014-09-17  new_confirmed      6\n",
       "2  2014-09-19  new_confirmed     12\n",
       "2  2014-09-21  new_confirmed     17\n",
       "2  2014-09-22  new_confirmed     14\n",
       "2  2014-09-23  new_confirmed     25\n",
       "2  2014-09-24  new_confirmed     20\n",
       "2  2014-09-26  new_confirmed     10\n",
       "2  2014-09-30  new_confirmed     11\n",
       "2  2014-10-01  new_confirmed      6"
      ]
     },
     "execution_count": 279,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PhilterList = ['new_confirmed', 'death_confirmed']\n",
    "frameSL = frameSL.loc[frameSL['Variable'].isin(PhilterList)]\n",
    "frameLI = frameLI.loc[frameLI['Variable'].isin(PhilterList)]\n",
    "frameGU = frameGU.loc[frameGU['Variable'].isin(PhilterList)]\n",
    "frameGU = frameGU[['Date', 'Variable', 'National']]\n",
    "frameGU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['population', 'new_noncase', 'new_suspected', 'new_probable',\n",
       "       'new_confirmed', 'cum_noncase', 'cum_suspected', 'cum_probable',\n",
       "       'cum_confirmed', 'death_suspected', 'death_probable',\n",
       "       'death_confirmed', 'cfr', 'etc_new_admission',\n",
       "       'etc_currently_admitted', 'etc_cum_admission', 'etc_new_deaths',\n",
       "       'etc_cum_deaths', 'etc_new_discharges', 'etc_cum_discharges',\n",
       "       'cum_contacts', 'cum_completed_contacts', 'contacts_followed',\n",
       "       'new_contacts', 'contacts_healthy', 'contacts_ill',\n",
       "       'contacts_not_seen', 'new_completed_contacts', 'percent_seen',\n",
       "       'total_lab_samples', 'repeat_samples', 'new_samples',\n",
       "       'new_positive', 'new_negative', 'pending', 'positive_corpse',\n",
       "       'negative_corpse'], dtype=object)"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameSL.Variable.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Specimens collected', 'Specimens pending for testing',\n",
       "       'Total specimens tested', 'Newly reported deaths',\n",
       "       'Total death/s in confirmed cases',\n",
       "       'Total death/s in probable cases',\n",
       "       'Total death/s in suspected cases',\n",
       "       'Total death/s in confirmed, probable, suspected cases',\n",
       "       'Case Fatality Rate (CFR) - Confirmed & Probable Cases',\n",
       "       'Newly reported contacts', 'Total contacts listed',\n",
       "       'Currently under follow-up', 'Contacts seen',\n",
       "       'Contacts who completed 21 day follow-up',\n",
       "       'Contacts lost to follow-up', 'New admissions',\n",
       "       'Total no. currently in Treatment Units', 'Total discharges',\n",
       "       'Cumulative admission/isolation', 'Newly Reported Cases in HCW',\n",
       "       'Cumulative cases among HCW', 'Newly Reported deaths in HCW',\n",
       "       'Cumulative deaths among HCW', 'New Case/s (Suspected)',\n",
       "       'New Case/s (Probable)', 'New case/s (confirmed)',\n",
       "       'Total suspected cases', 'Total probable cases',\n",
       "       'Total confirmed cases',\n",
       "       'Total Number of Confirmed Cases of Sierra Leonean Nationality',\n",
       "       'Total Number of Confirmed Cases of Guinean Nationality',\n",
       "       'Cumulative (confirmed + probable + suspected)',\n",
       "       'Total Case/s (Suspected)', 'Total Case/s (Probable)',\n",
       "       'Total case/s (confirmed)', 'Cumulative CFR',\n",
       "       'Cumulative (confirmed + probable + suspects)',\n",
       "       'Total death/s in confirmed,  probable, suspected cases',\n",
       "       'Cumulative confirmed, probable and suspected cases',\n",
       "       'Total death/s in confirmed, \\n probable, suspected cases',\n",
       "       'Case Fatality Rate (CFR) - \\n Confirmed & Probable Cases',\n",
       "       'Contacts who completed 21 day \\n follow-up',\n",
       "       'Total no. currently in Treatment \\n Units',\n",
       "       'Total Number of Confirmed Cases \\n of Sierra Leonean Nationality',\n",
       "       'Total Number of Confirmed Cases \\n of Guinean Nationality'], dtype=object)"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameLI.Variable.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "frameLI.index = frameLI.Variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'the label [new_confirmed] is not in the [index]'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_has_valid_type\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1433\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1434\u001b[0;31m                     \u001b[0merror\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1435\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36merror\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1428\u001b[0m                 raise KeyError(\"the label [%s] is not in the [%s]\" %\n\u001b[0;32m-> 1429\u001b[0;31m                                (key, self.obj._get_axis_name(axis)))\n\u001b[0m\u001b[1;32m   1430\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'the label [new_confirmed] is not in the [index]'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-164-db7d7f17c347>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mframeLI\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'new_confirmed'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1326\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1327\u001b[0m             \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_apply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1328\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1329\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1330\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1549\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1550\u001b[0m         \u001b[0;31m# fall thru to straight lookup\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1551\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_has_valid_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1552\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1553\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_has_valid_type\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1440\u001b[0m                 \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1441\u001b[0m             \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1442\u001b[0;31m                 \u001b[0merror\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1443\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1444\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36merror\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1427\u001b[0m                                     \"key\")\n\u001b[1;32m   1428\u001b[0m                 raise KeyError(\"the label [%s] is not in the [%s]\" %\n\u001b[0;32m-> 1429\u001b[0;31m                                (key, self.obj._get_axis_name(axis)))\n\u001b[0m\u001b[1;32m   1430\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1431\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'the label [new_confirmed] is not in the [index]'"
     ]
    }
   ],
   "source": [
    "frameLI.loc['new_confirmed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frameSL.index = frameSL.Date\n",
    "frameGU.index= frameGU.Date\n",
    "frameLI.index = frameLI.Variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "frameLI['Date'] = pd.to_datetime(frameLI.Date)\n",
    "frameLI.index = frameLI.Date\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ANSWER:\n",
    "We first import all the spreadsheets in separate dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "mbs = []\n",
    "for i in range(1,10):\n",
    "    temp = pd.read_excel(DATA_FOLDER+'/microbiome/MID'+str(i)+'.xls', sheetname='Sheet 1', header=None)\n",
    "    temp.columns = ['name', 'MID'+str(i)] # Using file name as column name for easier merging\n",
    "    mbs.append(temp)\n",
    "   \n",
    "    \n",
    "metadata = pd.read_excel(DATA_FOLDER+'/microbiome/metadata.xls', sheetname='Sheet1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we verify that the name column is a unique index for all the MID data and set this column as the index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mbs_ind = []\n",
    "doublons = False\n",
    "for i in range(1,10):\n",
    "    mbs_ind.append(mbs[i-1].set_index(['name']))\n",
    "    if mbs_ind[i-1].index.is_unique:\n",
    "        mbs[i-1] = mbs_ind[i-1]\n",
    "    else:\n",
    "        print(\"Table \" + i + \" contains doublons\")\n",
    "        doublons = True\n",
    "\n",
    "if not doublons:\n",
    "    print(\"Tables are indexed by the column name\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we can merge the MID tables using pandas' merge fucntion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import functools as ft\n",
    "mbs_merged = ft.reduce(lambda left,right: pd.merge(left,right,left_index=True, right_index=True, how='outer'), mbs)\n",
    "mbs_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now, we use the metadata file to have meaningfull headers for our data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we start be reordering the columns according to the metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "mbs_order = mbs_merged[metadata.BARCODE]\n",
    "mbs_order.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we split the table in two because the first column is 'EXTRACTION CONTROL' and does not have different types (tissue or stools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mbs_part1 = mbs_order.iloc[:, :1]\n",
    "mbs_part2 = mbs_order.iloc[:, 1:]\n",
    "mbs_part2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we create the 2-levels index according to the metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "header = pd.MultiIndex.from_product([['NEC1','control1','NEC2','control2'],\n",
    "                                     ['tissue','stool']],\n",
    "                                    names=['Test','Type'])\n",
    "mbs_part2.columns = header\n",
    "\n",
    "mbs_part1.columns = pd.MultiIndex.from_product([['EXTRACTION CONTROL'],['']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "merge again all the columns together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "mbs_nans = pd.merge(mbs_part1,mbs_part2,left_index=True, right_index=True, how='outer')\n",
    "mbs_nans.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first check the required invariants on index and dtypes, because it is easier before replacing NaNs by \"unknown\". Then we perform the replacement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from numpy import issubdtype, number\n",
    "print('Index are unique:', mbs_nans.index.is_unique)\n",
    "print('All entries are either a number or nan:', issubdtype(mbs_nans.dtypes.all(), number))\n",
    "mbs_final = mbs_nans.fillna(\"unknown\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We print the result:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mbs_final"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
